includes/polling/ports/os/smartax.inc.php

```diff
@@ -0,0 +1,73 @@
+<?php
+/**
+* smartax.inc.php
+*
+* This program is free software: you can redistribute it and/or modify
+* it under the terms of the GNU General Public License as published by
+* the Free Software Foundation, either version 3 of the License, or
+* (at your option) any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*
+* @package    LibreNMS
+* @link       http://librenms.org
+* @copyright  2020 Roman Tutkevich
+* @author     Roman Tutkevich <race.fdm@gmail.com>
+*/
+
+$inoctets = snmpwalk_cache_oid($device, '.1.3.6.1.4.1.2011.6.128.1.1.4.21.1.15', array());    //  hwGponOltEthernetStatisticReceivedBytes
```

Comment:
Cant you just walk `.1.3.6.1.4.1.2011.6.128.1.1.4.21.1` instead of 8 walks?

Reply:
@murrant I'm thinking If We need that code from @Negatifff because I updated my OLT's to latest R18 firmware and it started graphing GPON ports without this code. It just using IF-MIB for that.

Reply:
In the case of this device, I would not risk making one large walk instead of several small ones. In my practice, there have been cases when large and frequent (every 5 minutes) polls led to brakes when other information was sent on the SNMP.

Reply:
If IF-MIB is supported in a current firmware, it would be great to not have the os specific polling code.

Reply:
yes, in R18 if-mib counters change, maybe the vendor fixed it. but if it doesn’t work in old firmwares - maybe you should leave a survey of private mibs? In R18 counters in private mibs work, but maybe not everyone has the option of flashing the device.

Reply:
Maybe not everyone even wants data collected for GPON ports.
I would prefer not to have it if there is an alternative.

Reply:
do not take it for imposing, but in my opinion information about traffic on ports interests anyone who monitors a network device.

Reply:
But won't newer firmware versions poll duplicate data? Or am I misunderstanding here?

Reply:
Yes, you're right - new firmware will collect a duplicate. here you have to choose - or the data is everywhere or only on new firmware. perhaps need to add a survey of these branches, not in addition to existing polls, but instead of the existing poll.

Reply:
@Negatifff for example for me it didn't work to get traffic stats from GPON ports even with your code. I know that oid's are right from private MIB but I had an empty GPON ports graph s when I tested it. And I had firmwares from R0111 to R015, so I don't know if it's needed. Maybe it's working for you, you still don't answer to my question which concrete model do you have with which firmware. I can provide you newest firmware if you want. Or you can buy it as we did from guy fromgponsolutions. Trust me newest firmware giving us a lot of new options to provision services not only to get port statistics.

Reply:
I apologize, but I did not see a question about the firmware of my device. Besides the message of the conversation goes through the google translator for me. Maybe somewhere I misunderstood the meaning of the question asked to me.

I have about 30 stations. There are firmware R13 and R18. specifically the device on which I tested is Huawei SmartAX MA5600V800R013C10B039. and everywhere the branches given in the code work.

I made a small debug conclusion while waiting for the station with firmware p13:

smartax.inc.php
```
$inoctets = snmpwalk_cache_oid($device, '.1.3.6.1.4.1.2011.6.128.1.1.4.21.1.15', array());

print_r($inoctets["2011.6.128.1.1.4.21.1.15.4194356480"]);
sleep(1);
$inoctets = snmpwalk_cache_oid($device, '.1.3.6.1.4.1.2011.6.128.1.1.4.21.1.15', array());
print_r($inoctets["2011.6.128.1.1.4.21.1.15.4194356480"]);
sleep(1);
$inoctets = snmpwalk_cache_oid($device, '.1.3.6.1.4.1.2011.6.128.1.1.4.21.1.15', array());
print_r($inoctets["2011.6.128.1.1.4.21.1.15.4194356480"]);
die;

```

```
./poller.php -h 172.31.0.2
LibreNMS Poller
Starting polling run:

Hostname:    172.31.0.2
Device ID:   4
OS:          smartax
 (huawei)


#### Load poller module core ####
Uptime:  2 years 342 days 2 hours 45 minutes 23 seconds

>> Runtime for poller module 'core': 0.1078 seconds with 63408 bytes
>> SNMP: [2/0.10s] MySQL: [0/0.00s] RRD: [2/0.00s]
#### Unload poller module core ####

Module [ unix-agent ] disabled globally.

#### Load poller module os ####
Location: xxx
Hardware: SmartAX MA5600T
OS Version: MA5600V800R013C10B039

[...skipped...]

#### Load poller module ports ####
Caching Oids: Full ports polling ifDescr ifAdminStatus ifOperStatus ifLastChange ifType ifPhysAddress ifMtu ifInErrors ifOutErrors ifInDiscards ifOutDiscards dot3StatsDuplexStatusArray
(
    [enterprises] => 269025899247
)
Array
(
    [enterprises] => 269026474546
)
Array
(
    [enterprises] => 269026636055
)
```

In addition, as I said, I draw graphs of all ports of stations in Cricket from these snmp branches.

I understand that a new firmware is almost always something new and improved :). Difficulties can cause 2 points: 1) the appropriateness of the firmware, if everything works. to reflash a station that is in production, only to reflash - why? 2) the one who monitors the stations may simply not be able to reflash them. for example lack of rights.

Reply:
I'm not sure what the conclusion is here

Reply:
@Jellyfrog to use @Negatifff code to get port stats from smartax MIB to have get that data with older firmware as well with newest firmwares :)

Reply:
If that is the conclusion, then the code should do one walk, and check for output.  If there is no output, skip the other walks otherwise we have to wait for each walk to fail which can greatly increase polling time.

Reply:
@murrant So ok I will split this to another PR I really want to have these sensors alredy implemented, so it will be ok ? Then we can merge this PR without port code from @Negatifff ? And we will try to improve the code for ports polling

Reply:
Do I understand correctly that in this case devices with older firmware will remain overboard and will not draw graphs?

Reply:
@Negatifff we need to improve code as @murrant suggested and then we can add that in another PR for example

Reply:
guys, I'm serious - you don't need to read a lot of information from these stations with one request. this can lead to problems with ssn access to these devices. besides, at one time you will have to poll the tree 1.3.6.1.4.1.2011.6.128.1.1.4.21.1, and there are 31 branches. what for? Why get 75% of unnecessary information?

Reply:
very sorry, I answered a slightly different request.
do you want to check after each survey? what for?

Reply:
Yes, easy merge without the port code.